File: apps/android/app/src/main/java/hub/core/android/chat/HubFiles.kt
package hub.core.android.chat

import hub.core.android.data.HubError
import hub.core.android.data.apiBase
import hub.core.android.data.hubCall
import hub.core.client.api.MetaApi
import hub.core.client.api.SessionsApi
import hub.core.client.model.SessionFile
import hub.core.client.model.SessionsCreateFileStreamRequest
import java.io.File
import java.io.InterruptedIOException
import java.security.MessageDigest
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType
import okhttp3.OkHttpClient
import okhttp3.ResponseBody
import okio.Buffer
import okio.BufferedSource
import okio.ForwardingSource
import okio.buffer

/*
 * Opening the files of a conversation on the phone (docs/changes/2026-09-26-twuijri-mobile-open-files.md).
 * Everything here is pure or talks to the hub through the generated client only, so the rules are
 * tested on the JVM (HubFilesTest): which way a file opens, which file a link in a reply names, and
 * the download itself (bearer header, progress, cancel).
 */

/** A file of the hub a message carries or a reply's link names. */
sealed interface HubFile {
    val name: String
    val mime: String?
    val sizeBytes: Long?

    /** The folder it is kept under in the phone's cache. */
    val cacheKey: String

    /** An attachment: its bytes never change (`sessions.downloadAttachment`). */
    data class Attachment(val id: String, override val name: String, override val mime: String? = null, override val sizeBytes: Long? = null) : HubFile {
        override val cacheKey: String get() = id
    }

    /**
     * A file of the conversation's working folder (`sessions.readFile`): it may change, so the key
     * carries its size and time when the hub gave them.
     */
    data class Working(
        val sessionId: String,
        val path: String,
        override val name: String,
        override val mime: String? = null,
        override val sizeBytes: Long? = null,
        val modified: String? = null,
    ) : HubFile {
        override val cacheKey: String get() = "file-" + sha1("$sessionId\u0000$path\u0000$sizeBytes\u0000$modified").take(24)
    }

    companion object {
        fun of(attachment: ChatAttachment): Attachment? = attachment.attachmentId?.let {
            Attachment(it, attachment.name ?: it, attachment.mime, attachment.sizeBytes)
        }

        /** An entry of `sessions.listFiles`: its attachment when it is one, else the folder's file. */
        fun of(file: SessionFile, sessionId: String): HubFile = file.attachmentId?.let { Attachment(it, file.name, file.mime, file.sizeBytes) }
            ?: Working(sessionId, file.path ?: file.name, file.name, file.mime, file.sizeBytes, file.modifiedAt?.toString())
    }
}

private fun sha1(text: String): String =
    MessageDigest.getInstance("SHA-1").digest(text.toByteArray()).joinToString("") { "%02x".format(it) }

/** How a file opens: drawn by the app, in the phone's viewer or player, or handed to the share sheet. */
enum class FileOpen { PICTURE, VIEWER, MEDIA, SHARE }

object FileKinds {
    /** Pictures the app decodes itself (an SVG is not one: it opens in the viewer). */
    private val pictures = setOf("image/jpeg", "image/png", "image/gif", "image/webp", "image/heic", "image/heif", "image/bmp")

    private val byExtension = mapOf(
        "jpg" to "image/jpeg", "jpeg" to "image/jpeg", "png" to "image/png", "gif" to "image/gif", "webp" to "image/webp",
        "heic" to "image/heic", "heif" to "image/heif", "bmp" to "image/bmp", "svg" to "image/svg+xml",
        "pdf" to "application/pdf", "txt" to "text/plain", "md" to "text/markdown", "csv" to "text/csv",
        "json" to "application/json", "html" to "text/html", "htm" to "text/html", "xml" to "text/xml", "log" to "text/plain",
        "mp3" to "audio/mpeg", "m4a" to "audio/mp4", "aac" to "audio/aac", "wav" to "audio/wav", "ogg" to "audio/ogg",
        "oga" to "audio/ogg", "opus" to "audio/ogg", "flac" to "audio/flac", "weba" to "audio/webm",
        "mp4" to "video/mp4", "m4v" to "video/mp4", "mov" to "video/quicktime", "webm" to "video/webm", "mkv" to "video/x-matroska",
        "doc" to "application/msword", "docx" to "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "xls" to "application/vnd.ms-excel", "xlsx" to "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "ppt" to "application/vnd.ms-powerpoint", "pptx" to "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "epub" to "application/epub+zip", "rtf" to "application/rtf",
    )

    /** Opened in the phone's own viewer: documents and text. Anything else is shared or saved. */
    private val documents = setOf(
        "application/pdf", "application/json", "application/rtf", "application/epub+zip", "image/svg+xml",
        "application/msword", "application/vnd.ms-excel", "application/vnd.ms-powerpoint",
    )

    /** The file's type: the hub's, unless it says nothing (`application/octet-stream`); then its name's. */
    fun mimeOf(name: String, mime: String?): String {
        val given = mime?.substringBefore(';')?.trim()?.lowercase()
        if (!given.isNullOrEmpty() && given != "application/octet-stream") return given
        return byExtension[name.substringAfterLast('.', "").lowercase()] ?: "application/octet-stream"
    }

    fun openAs(name: String, mime: String?): FileOpen {
        val type = mimeOf(name, mime)
        return when {
            type in pictures -> FileOpen.PICTURE
            type.startsWith("audio/") || type.startsWith("video/") -> FileOpen.MEDIA
            type.startsWith("text/") || type in documents || type.startsWith("application/vnd.openxmlformats-officedocument.") -> FileOpen.VIEWER
            else -> FileOpen.SHARE
        }
    }

    fun openAs(file: HubFile): FileOpen = openAs(file.name, file.mime)
}

/**
 * Which file of the hub a link in a reply names (web: `Markdown.tsx`, decision §48): the address of
 * one of the reply's attachments (relative or on the hub's own origin), or a word that is one of the
 * conversation's files — its path, a path ending in it, or its name when one file has it (the reply's
 * own files first). Any other link is an ordinary link and opens in the browser.
 */
object FileLinks {
    /** The address a hub serves an attachment's bytes at, as the generated client builds it. */
    fun contentPath(attachmentId: String): String =
        MetaApi.defaultBasePath + SessionsApi(apiBase("")).sessionsDownloadAttachmentRequestConfig("", attachmentId, null).path

    /** The link as a word: decoded, without a query or fragment; null when it points off the hub. */
    fun wordOf(href: String, hub: String): String? {
        val raw = href.trim()
        if (raw.isEmpty() || raw.startsWith("#")) return null
        val scheme = Regex("^([a-zA-Z][a-zA-Z0-9+.-]*):").find(raw)?.groupValues?.get(1)?.lowercase()
        val path = when {
            scheme == "http" || scheme == "https" -> {
                val url = raw.toHttpUrlOrNull() ?: return null
                val home = hub.toHttpUrlOrNull() ?: return null
                if (url.scheme != home.scheme || url.host != home.host || url.port != home.port) return null
                url.encodedPath
            }
            // What agents write for a file they made: `sandbox:/…`, `file:///…`.
            scheme == "sandbox" || scheme == "file" -> raw.substringAfter(':').trimStart('/').let { "/$it" }
            scheme != null -> return null
            raw.startsWith("//") -> return null
            else -> raw
        }
        val bare = path.substringBefore('#').substringBefore('?')
        return runCatching { java.net.URLDecoder.decode(bare.replace("+", "%2B"), "UTF-8") }.getOrDefault(bare).ifBlank { null }
    }

    /** The file [href] names, or null for an ordinary link. [files] is the conversation's list, when read. */
    fun resolve(href: String, hub: String, own: List<ChatAttachment>, files: List<SessionFile>? = null, sessionId: String? = null): HubFile? {
        val word = wordOf(href, hub) ?: return null
        // The address of an attachment the reply carries (the block's own `url`, or where the hub serves it).
        own.forEach { a ->
            val id = a.attachmentId ?: return@forEach
            if (word == a.url || word == contentPath(id)) return HubFile.of(a)
        }
        val known = files.orEmpty()
        known.firstOrNull { it.attachmentId != null && word == contentPath(it.attachmentId!!) }?.let { return HubFile.of(it, sessionId ?: "") }
        return forWord(word, own, known, sessionId)
    }

    private fun forWord(mention: String, own: List<ChatAttachment>, files: List<SessionFile>, sessionId: String?): HubFile? {
        val word = mention.trim().removePrefix("./")
        if (word.isEmpty() || word.length > 4096 || word.any { it.isWhitespace() && it != ' ' }) return null
        val base = word.substringAfterLast('/')
        // The reply's own file of that name first: `flying_cat.png` means the one it carries.
        own.filter { it.attachmentId != null && it.name == base }.singleOrNull()?.let { return HubFile.of(it) }
        if (sessionId == null) return null
        val onDisk = files.filter { it.path != null }
        onDisk.firstOrNull { it.path == word || word.endsWith("/${it.path}") }?.let { return HubFile.of(it, sessionId) }
        return files.filter { it.name == base }.singleOrNull()?.let { HubFile.of(it, sessionId) }
    }
}

/** How far a download is: [read] bytes of [total] (null when the hub did not say). */
data class Progress(val read: Long, val total: Long?) {
    val fraction: Float? get() = total?.takeIf { it > 0 }?.let { (read.toFloat() / it).coerceIn(0f, 1f) }
}

/**
 * Fetches a file of the hub into the phone's cache through the generated client (the bearer goes in
 * the header, never in the address), reporting progress and stopping when its coroutine is cancelled.
 * A file is written beside its place and moved in only once whole, so a download cut short is never
 * mistaken for the file.
 */
class HubFileFetcher(private val client: OkHttpClient, private val root: File) {
    /** Where [file] is kept: a folder per file, under its own (safe) name so a viewer knows its kind. */
    fun place(file: HubFile): File = File(File(root, file.cacheKey), safeName(file.name, file.cacheKey))

    /** The file when it is already here, whole. */
    fun cached(file: HubFile): File? = place(file).takeIf { it.isFile && (file.sizeBytes == null || it.length() == file.sizeBytes) }

    suspend fun fetch(hub: String, profile: String, file: HubFile, onProgress: (Progress) -> Unit = {}): File {
        cached(file)?.let { return it }
        val job = currentCoroutineContext()[Job]
        val counting = client.newBuilder().addNetworkInterceptor { chain ->
            val response = chain.proceed(chain.request())
            val body = response.body
            if (!response.isSuccessful || body == null) response
            else response.newBuilder().body(CountingBody(body, file.sizeBytes, onProgress) { job?.isActive != false }).build()
        }.build()
        val sessions = SessionsApi(apiBase(hub), counting)
        val downloaded = try {
            hubCall {
                when (file) {
                    is HubFile.Attachment -> sessions.sessionsDownloadAttachment(xHubProfile = profile, attachmentId = file.id)
                    is HubFile.Working -> sessions.sessionsReadFile(xHubProfile = profile, sessionId = file.sessionId, path = file.path, download = true)
                }
            }.getOrThrow()
        } catch (e: HubError) {
            if (job?.isActive == false) throw CancellationException("download cancelled")
            throw e
        }
        return withContext(Dispatchers.IO) {
            val target = place(file)
            target.parentFile?.mkdirs()
            val part = File(target.parentFile, target.name + ".part")
            try {
                downloaded.copyTo(part, overwrite = true)
                if (!part.renameTo(target)) {
                    part.copyTo(target, overwrite = true)
                }
                target
            } finally {
                part.delete()
                downloaded.delete()
            }
        }
    }

    /**
     * An address the phone's player can stream an audio or video file from, with byte ranges and no
     * bearer: the contract's one-hour ticket for this one file (DECISIONS §90, §98), resolved against
     * the hub's origin.
     */
    suspend fun streamAddress(hub: String, profile: String, file: HubFile): String {
        val sessions = SessionsApi(apiBase(hub), client)
        val ticket = hubCall {
            when (file) {
                is HubFile.Attachment -> sessions.sessionsCreateAttachmentStream(xHubProfile = profile, attachmentId = file.id)
                is HubFile.Working -> sessions.sessionsCreateFileStream(profile, file.sessionId, SessionsCreateFileStreamRequest(path = file.path))
            }
        }.getOrThrow()
        return resolve(hub, ticket.url) ?: throw HubError(-1, "bad_address", ticket.url)
    }

    companion object {
        /** A name that stays inside its folder: no separators, never empty. */
        fun safeName(name: String, fallback: String): String =
            name.replace('/', '_').replace('\\', '_').replace(':', '_').trim().ifBlank { fallback }.takeLast(200)

        /** [url] (relative to the hub, or whole) as a whole address on the hub. */
        fun resolve(hub: String, url: String): String? = hub.toHttpUrlOrNull()?.resolve(url)?.toString()
    }
}

/** A body that tells how much of it has been read, and stops reading once [active] says so. */
private class CountingBody(
    private val body: ResponseBody,
    private val size: Long?,
    private val onProgress: (Progress) -> Unit,
    private val active: () -> Boolean,
) : ResponseBody() {
    override fun contentType(): MediaType? = body.contentType()
    override fun contentLength(): Long = body.contentLength()
    private val counted: BufferedSource by lazy {
        val total = body.contentLength().takeIf { it >= 0 } ?: size
        object : ForwardingSource(body.source()) {
            var read = 0L
            override fun read(sink: Buffer, byteCount: Long): Long {
                if (!active()) throw InterruptedIOException("download cancelled")
                val n = super.read(sink, byteCount)
                if (n > 0) read += n
                onProgress(Progress(read, total))
                return n
            }
        }.buffer()
    }
    override fun source(): BufferedSource = counted
}

/**
 * The downloads of a conversation's files, one per file, for the app's life: an inline picture and
 * a tap on it share one download, leaving the chat does not cut a large one short, and a person can
 * cancel it. [hub] is the signed-in hub's address; [profile] the file's conversation's profile.
 */
class FileDownloads(
    val fetcher: HubFileFetcher,
    private val hub: () -> String?,
    private val scope: kotlinx.coroutines.CoroutineScope,
) {
    sealed interface State {
        data object Idle : State
        data class Loading(val progress: Progress?) : State
        data class Ready(val file: File) : State

        /** [error] is the hub's own words when it sent some; null when the phone could not reach it. */
        data class Failed(val error: HubError) : State
    }

    private val _states = kotlinx.coroutines.flow.MutableStateFlow<Map<String, State>>(emptyMap())
    val states: kotlinx.coroutines.flow.StateFlow<Map<String, State>> = _states
    private val jobs = java.util.concurrent.ConcurrentHashMap<String, Job>()

    fun state(file: HubFile): State = _states.value[file.cacheKey]
        ?: fetcher.cached(file)?.let { State.Ready(it) }
        ?: State.Idle

    private fun set(key: String, state: State?) = _states.update { if (state == null) it - key else it + (key to state) }

    /** Starts fetching [file] unless it is here or on its way; the state says when it is ready. */
    fun start(file: HubFile, profile: String) {
        val key = file.cacheKey
        val now = state(file)
        if (now is State.Ready && now.file.isFile) return set(key, now)
        if (jobs[key]?.isActive == true) return
        val address = hub() ?: return set(key, State.Failed(HubError(401, "unauthorized", null)))
        set(key, State.Loading(null))
        var last = 0L
        jobs[key] = scope.launch {
            try {
                val local = fetcher.fetch(address, profile, file) { p ->
                    // At most every 64 KB, or the end: enough for a bar, not a storm of states.
                    if (p.read - last >= 65_536 || p.read == p.total) {
                        last = p.read
                        set(key, State.Loading(p))
                    }
                }
                set(key, State.Ready(local))
            } catch (e: CancellationException) {
                set(key, null)
                throw e
            } catch (e: HubError) {
                set(key, State.Failed(e))
            } catch (e: Throwable) {
                set(key, State.Failed(HubError.from(e)))
            } finally {
                jobs.remove(key)
            }
        }
    }

    /** Stops a download; the file goes back to not fetched. */
    fun cancel(file: HubFile) {
        jobs.remove(file.cacheKey)?.cancel()
        set(file.cacheKey, null)
    }

    /** The player's address for an audio or video file (a one-hour ticket), or why there is none. */
    suspend fun streamAddress(file: HubFile, profile: String): Result<String> {
        val address = hub() ?: return Result.failure(HubError(401, "unauthorized", null))
        return runCatching { fetcher.streamAddress(address, profile, file) }
    }
}

File: apps/ios/CoreHub/Chat/HubFiles.swift
// Opening the files of a conversation on the phone (docs/changes/2026-09-26-twuijri-mobile-open-files.md):
// which way a file opens, which file a link in a reply names, and the download itself — through the
// generated client only, the bearer in the header (HubAPI), with progress and cancel. The rules are
// pure and tested (HubFilesTests); the views are in Attachments.swift.
import CoreHubClient
import CryptoKit
import Foundation

/// A file of the hub a message carries or a reply's link names.
enum HubFile: Hashable {
    /// An attachment: its bytes never change (`sessions.downloadAttachment`).
    case attachment(id: String, name: String, mime: String?, size: Int?)
    /// A file of the conversation's working folder (`sessions.readFile`); it may change.
    case working(sessionID: String, path: String, name: String, mime: String?, size: Int?, modified: Date?)

    var name: String {
        switch self {
        case .attachment(_, let name, _, _), .working(_, _, let name, _, _, _): return name
        }
    }

    var mime: String? {
        switch self {
        case .attachment(_, _, let mime, _), .working(_, _, _, let mime, _, _): return mime
        }
    }

    var size: Int? {
        switch self {
        case .attachment(_, _, _, let size), .working(_, _, _, _, let size, _): return size
        }
    }

    /// The folder it is kept under in the phone's caches: the attachment's id, or a digest of the
    /// working file's place, size and time (a changed file is fetched again).
    var cacheKey: String {
        switch self {
        case .attachment(let id, _, _, _):
            return id
        case .working(let session, let path, _, _, let size, let modified):
            let seed = "\(session)\u{0}\(path)\u{0}\(size.map(String.init) ?? "")\u{0}\(modified.map { String($0.timeIntervalSince1970) } ?? "")"
            let digest = Insecure.SHA1.hash(data: Data(seed.utf8)).map { String(format: "%02x", $0) }.joined()
            return "file-" + digest.prefix(24)
        }
    }

    /// An entry of `sessions.listFiles`: its attachment when it is one, else the folder's file.
    static func of(_ file: SessionFile, sessionID: String) -> HubFile {
        if let id = file.attachmentId { return .attachment(id: id, name: file.name, mime: file.mime, size: file.sizeBytes) }
        return .working(sessionID: sessionID, path: file.path ?? file.name, name: file.name, mime: file.mime, size: file.sizeBytes, modified: file.modifiedAt)
    }
}

/// How a file opens: drawn by the app, in the system's viewer (Quick Look), played, or shared.
enum FileOpen: Equatable {
    case picture, viewer, media, share
}

enum FileKinds {
    /// Pictures the app draws itself (an SVG is not one: it opens in the viewer).
    private static let pictures: Set<String> = ["image/jpeg", "image/png", "image/gif", "image/webp", "image/heic", "image/heif", "image/bmp"]

    private static let byExtension: [String: String] = [
        "jpg": "image/jpeg", "jpeg": "image/jpeg", "png": "image/png", "gif": "image/gif", "webp": "image/webp",
        "heic": "image/heic", "heif": "image/heif", "bmp": "image/bmp", "svg": "image/svg+xml",
        "pdf": "application/pdf", "txt": "text/plain", "md": "text/markdown", "csv": "text/csv",
        "json": "application/json", "html": "text/html", "htm": "text/html", "xml": "text/xml", "log": "text/plain",
        "mp3": "audio/mpeg", "m4a": "audio/mp4", "aac": "audio/aac", "wav": "audio/wav", "ogg": "audio/ogg",
        "oga": "audio/ogg", "opus": "audio/ogg", "flac": "audio/flac", "weba": "audio/webm",
        "mp4": "video/mp4", "m4v": "video/mp4", "mov": "video/quicktime", "webm": "video/webm", "mkv": "video/x-matroska",
        "doc": "application/msword", "docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        "xls": "application/vnd.ms-excel", "xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        "ppt": "application/vnd.ms-powerpoint", "pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation",
        "epub": "application/epub+zip", "rtf": "application/rtf",
    ]

    /// Opened in the system's viewer: documents and text. Anything else is shared or saved.
    private static let documents: Set<String> = [
        "application/pdf", "application/json", "application/rtf", "application/epub+zip", "image/svg+xml",
        "application/msword", "application/vnd.ms-excel", "application/vnd.ms-powerpoint",
    ]

    /// The file's type: the hub's, unless it says nothing (`application/octet-stream`); then its name's.
    static func mime(of name: String, _ mime: String?) -> String {
        let given = (mime ?? "").split(separator: ";").first.map { $0.trimmingCharacters(in: .whitespaces).lowercased() } ?? ""
        if !given.isEmpty, given != "application/octet-stream" { return given }
        let ext = (name as NSString).pathExtension.lowercased()
        return byExtension[ext] ?? "application/octet-stream"
    }

    static func openAs(_ name: String, _ mime: String?) -> FileOpen {
        let type = self.mime(of: name, mime)
        if pictures.contains(type) { return .picture }
        if type.hasPrefix("audio/") || type.hasPrefix("video/") { return .media }
        if type.hasPrefix("text/") || documents.contains(type) || type.hasPrefix("application/vnd.openxmlformats-officedocument.") { return .viewer }
        return .share
    }

    static func openAs(_ file: HubFile) -> FileOpen { openAs(file.name, file.mime) }
}

/// Which file of the hub a link in a reply names (web: `Markdown.tsx`, decision §48): the address of
/// one of the reply's attachments (relative, or on the hub's own origin), or a word that is one of
/// the conversation's files — its path, a path ending in it, or its name when one file has it (the
/// reply's own files first). Any other link is an ordinary link.
enum FileLinks {
    /// The address a hub serves an attachment's bytes at, as the generated client builds it.
    static func contentPath(_ attachmentID: String) -> String {
        let config = CoreHubClientAPIConfiguration(basePath: CoreHubClientAPIConfiguration().basePath)
        return SessionsAPI.sessionsDownloadAttachmentWithRequestBuilder(xHubProfile: "", attachmentId: attachmentID, apiConfiguration: config).URLString
    }

    /// The link as a word: decoded, without a query or fragment; nil when it points off the hub.
    static func word(of href: String, hub: URL?) -> String? {
        let raw = href.trimmingCharacters(in: .whitespacesAndNewlines)
        if raw.isEmpty || raw.hasPrefix("#") || raw.hasPrefix("//") { return nil }
        var path = raw
        if let colon = raw.firstIndex(of: ":"),
           raw[raw.startIndex..<colon].range(of: #"^[A-Za-z][A-Za-z0-9+.\-]*$"#, options: .regularExpression) != nil {
            let scheme = raw[raw.startIndex..<colon].lowercased()
            switch scheme {
            case "http", "https":
                guard let url = URLComponents(string: raw), let hub, let home = URLComponents(url: hub, resolvingAgainstBaseURL: false),
                      url.scheme?.lowercased() == home.scheme?.lowercased(), url.host?.lowercased() == home.host?.lowercased(),
                      (url.port ?? defaultPort(url.scheme)) == (home.port ?? defaultPort(home.scheme)) else { return nil }
                path = url.percentEncodedPath
            case "sandbox", "file":
                // What agents write for a file they made: `sandbox:/…`, `file:///…`.
                let rest = raw[raw.index(after: colon)...].drop { $0 == "/" }
                path = "/" + rest
            default:
                return nil
            }
        }
        if let cut = path.firstIndex(where: { $0 == "?" || $0 == "#" }) { path = String(path[..<cut]) }
        let decoded = path.removingPercentEncoding ?? path
        return decoded.isEmpty ? nil : decoded
    }

    private static func defaultPort(_ scheme: String?) -> Int? {
        switch scheme?.lowercased() {
        case "http": return 80
        case "https": return 443
        default: return nil
        }
    }

    /// The file `href` names, or nil for an ordinary link. `files` is the conversation's list, when read.
    static func resolve(_ href: String, hub: URL?, own: [MessageAttachments.File], files: [SessionFile]? = nil, sessionID: String? = nil) -> HubFile? {
        guard let word = word(of: href, hub: hub) else { return nil }
        for file in own where word == file.url || word == contentPath(file.attachmentID) {
            return file.hubFile
        }
        let known = files ?? []
        if let entry = known.first(where: { $0.attachmentId.map { word == contentPath($0) } ?? false }) {
            return HubFile.of(entry, sessionID: sessionID ?? "")
        }
        var mention = word.trimmingCharacters(in: .whitespaces)
        if mention.hasPrefix("./") { mention.removeFirst(2) }
        guard !mention.isEmpty, mention.count <= 4096, !mention.contains(where: { $0.isNewline || $0 == "\t" }) else { return nil }
        let base = (mention as NSString).lastPathComponent
        // The reply's own file of that name first: `flying_cat.png` means the one it carries.
        let mine = own.filter { $0.name == base }
        if mine.count == 1 { return mine[0].hubFile }
        guard let sessionID else { return nil }
        if let exact = known.first(where: { entry in entry.path.map { mention == $0 || mention.hasSuffix("/" + $0) } ?? false }) {
            return HubFile.of(exact, sessionID: sessionID)
        }
        let named = known.filter { $0.name == base }
        return named.count == 1 ? HubFile.of(named[0], sessionID: sessionID) : nil
    }
}

/// Fetching one file of the hub, one attempt with a given configuration of the generated client
/// (HubAPI.call adds the bearer and retries once after a refresh).
enum HubFileFetcher {
    /// Where a file is kept: a folder per file, under its own (safe) name so Quick Look knows its kind.
    static func place(_ file: HubFile, in root: URL) -> URL {
        root.appendingPathComponent(file.cacheKey, isDirectory: true).appendingPathComponent(safeName(file.name, fallback: file.cacheKey))
    }

    /// A name that stays inside its folder: no separators, never empty.
    static func safeName(_ name: String, fallback: String) -> String {
        let safe = name.replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "\\", with: "_").replacingOccurrences(of: ":", with: "_")
            .trimmingCharacters(in: .whitespaces)
        return safe.isEmpty ? fallback : String(safe.suffix(200))
    }

    /// The file when it is already here, whole.
    static func cached(_ file: HubFile, in root: URL) -> URL? {
        let url = place(file, in: root)
        guard let attributes = try? FileManager.default.attributesOfItem(atPath: url.path) else { return nil }
        if let size = file.size, let stored = attributes[.size] as? Int, stored != size { return nil }
        return url
    }

    /// Downloads `file` with `config`; `progress` hears the fraction as bytes arrive (nil while unknown).
    static func download(_ file: HubFile, profile: String, config: CoreHubClientAPIConfiguration, progress: @escaping @Sendable (Double?) -> Void) async throws -> URL {
        let builder: RequestBuilder<URL>
        switch file {
        case .attachment(let id, _, _, _):
            builder = SessionsAPI.sessionsDownloadAttachmentWithRequestBuilder(xHubProfile: profile, attachmentId: id, apiConfiguration: config)
        case .working(let session, let path, _, _, _, _):
            builder = SessionsAPI.sessionsReadFileWithRequestBuilder(xHubProfile: profile, sessionId: session, path: path, download: true, apiConfiguration: config)
        }
        let watcher = ProgressWatcher(progress)
        builder.onProgressReady = { watcher.watch($0) }
        defer { watcher.stop() }
        return try await builder.execute().body
    }

    /// Moves what the generated client wrote into the file's place, replacing an older copy.
    static func keep(_ downloaded: URL, as file: HubFile, in root: URL) throws -> URL {
        let target = place(file, in: root)
        let manager = FileManager.default
        try manager.createDirectory(at: target.deletingLastPathComponent(), withIntermediateDirectories: true)
        try? manager.removeItem(at: target)
        try manager.moveItem(at: downloaded, to: target)
        return target
    }

    /// An address the system's player streams an audio or video file from, with byte ranges and
    /// no bearer: the contract's one-hour ticket for this one file (DECISIONS §90, §98), resolved
    /// against the hub.
    static func streamAddress(_ file: HubFile, profile: String, hub: URL, config: CoreHubClientAPIConfiguration) async throws -> URL {
        let ticket: AttachmentStream
        switch file {
        case .attachment(let id, _, _, _):
            ticket = try await SessionsAPI.sessionsCreateAttachmentStream(xHubProfile: profile, attachmentId: id, apiConfiguration: config)
        case .working(let session, let path, _, _, _, _):
            ticket = try await SessionsAPI.sessionsCreateFileStream(
                xHubProfile: profile, sessionId: session, sessionsCreateFileStreamRequest: SessionsCreateFileStreamRequest(path: path), apiConfiguration: config
            )
        }
        guard let address = resolve(ticket.url, against: hub) else { throw HubFailure(kind: .decode, status: 200, code: nil, message: nil, operationID: nil, requestID: nil, detail: ticket.url) }
        return address
    }

    /// `url` (relative to the hub, or whole) as a whole address on the hub.
    static func resolve(_ url: String, against hub: URL) -> URL? {
        URL(string: url, relativeTo: hub)?.absoluteURL
    }
}

/// Reports a download's `Progress` as a fraction, until stopped.
final class ProgressWatcher: @unchecked Sendable {
    private let report: @Sendable (Double?) -> Void
    private var observation: NSKeyValueObservation?
    private let lock = NSLock()

    init(_ report: @escaping @Sendable (Double?) -> Void) {
        self.report = report
    }

    func watch(_ progress: Progress) {
        let report = self.report
        let observation = progress.observe(\.fractionCompleted, options: [.new]) { progress, _ in
            report(progress.totalUnitCount > 0 ? progress.fractionCompleted : nil)
        }
        lock.lock()
        self.observation = observation
        lock.unlock()
    }

    func stop() {
        lock.lock()
        observation?.invalidate()
        observation = nil
        lock.unlock()
    }
}

/// The downloads of a conversation's files, one per file, for the app's life: an inline picture
/// and a tap on it share one download, leaving the chat does not cut a large one short, and a
/// person can cancel it.
@MainActor
@Observable
final class FileDownloads {
    enum State: Equatable {
        case idle
        /// The fraction arrived, nil while the hub has not said how large the file is.
        case loading(Double?)
        case ready(URL)
        case failed(HubFailure)
    }

    static let shared = FileDownloads()

    private(set) var states: [String: State] = [:]
    @ObservationIgnored private var tasks: [String: Task<Void, Never>] = [:]

    private var root: URL {
        FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask)[0].appendingPathComponent("attachments", isDirectory: true)
    }

    func state(_ file: HubFile) -> State {
        if let state = states[file.cacheKey] { return state }
        if let url = HubFileFetcher.cached(file, in: root) { return .ready(url) }
        return .idle
    }

    /// Starts fetching `file` unless it is here or on its way.
    func start(_ file: HubFile, profile: String, app: AppModel) {
        let key = file.cacheKey
        if case .ready(let url) = state(file), FileManager.default.fileExists(atPath: url.path) {
            states[key] = .ready(url)
            return
        }
        if tasks[key] != nil { return }
        states[key] = .loading(nil)
        let root = self.root
        let scope = profile.isEmpty ? app.currentProfile : profile
        tasks[key] = Task {
            do {
                let downloaded = try await app.api.call { config in
                    try await HubFileFetcher.download(file, profile: scope, config: config) { fraction in
                        Task { @MainActor in
                            guard case .loading = FileDownloads.shared.states[key] else { return }
                            FileDownloads.shared.states[key] = .loading(fraction)
                        }
                    }
                }
                let kept = try HubFileFetcher.keep(downloaded, as: file, in: root)
                self.finish(key, .ready(kept))
            } catch {
                if Task.isCancelled { self.finish(key, nil) } else { self.finish(key, .failed(HubFailure(error))) }
            }
        }
    }

    private func finish(_ key: String, _ state: State?) {
        tasks[key] = nil
        states[key] = state
    }

    /// Stops a download; the file goes back to not fetched.
    func cancel(_ file: HubFile) {
        tasks.removeValue(forKey: file.cacheKey)?.cancel()
        states[file.cacheKey] = nil
    }

    /// The player's address for an audio or video file (a one-hour ticket).
    func streamAddress(_ file: HubFile, profile: String, app: AppModel) async throws -> URL {
        guard let hub = app.credentials?.hubURL else { throw HubFailure.signedOut }
        let scope = profile.isEmpty ? app.currentProfile : profile
        return try await app.api.call { config in
            try await HubFileFetcher.streamAddress(file, profile: scope, hub: hub, config: config)
        }
    }
}
